flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/DeduplicateFunctionHelper.java

```diff
@@ -35,26 +35,31 @@
 	 *
 	 * @param currentRow latest row received by deduplicate function
 	 * @param generateUpdateBefore whether need to send UPDATE_BEFORE message for updates
-	 * @param state state of function
+	 * @param state state of function, null if generateUpdateBefore is false
 	 * @param out underlying collector
 	 */
 	static void processLastRow(
 			BaseRow currentRow,
 			boolean generateUpdateBefore,
 			ValueState<BaseRow> state,
 			Collector<BaseRow> out) throws Exception {
-		// Check message should be accumulate
-		Preconditions.checkArgument(BaseRowUtil.isAccumulateMsg(currentRow));
-		if (generateUpdateBefore) {
-			// state stores complete row if generateUpdateBefore is true
-			BaseRow preRow = state.value();
-			state.update(currentRow);
-			if (preRow != null) {
-				preRow.setHeader(BaseRowUtil.RETRACT_MSG);
+		// check message should be insert only.
+		Preconditions.checkArgument(currentRow.getRowKind() == RowKind.INSERT);
+		BaseRow preRow = state.value();
+		state.update(currentRow);
```

Comment:
After offline disscussion, I will add an optimization configuration to always send AFTER instead of INSERT for first row. This can avoid state accessing but is disabled by default.